File: js/src/devices.ts
/**
 * `devices` resource client.
 *
 * Mounted on `Nitroping` as `np.devices`. Wraps `GET /api/v1/devices`,
 * `POST /api/v1/devices`, `PUT /api/v1/devices/:id`, `DELETE
 * /api/v1/devices/:id`, and `DELETE /api/v1/devices` (by token).
 */

import type { HttpClient } from "./http"
import type {
  ListDevicesQuery,
  ListDevicesResponse,
  RegisterDeviceRequest,
  RegisterDeviceResponse,
  UpdateDeviceRequest,
  UpdateDeviceResponse,
} from "./types"

interface RawDeviceSummary {
  id: string
  user_id: string | null
  platform: RegisterDeviceRequest["platform"]
  status: "active" | "inactive"
  tags: string[]
  timezone: string | null
  apns_environment: "sandbox" | "production" | null
  last_seen_at: string | null
  inserted_at: string
}

interface RawListDevicesResponse {
  data: RawDeviceSummary[]
  total: number
}

export class DevicesClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Register (or update) a device.
   *
   * Uses `/api/v1/public/devices` when the client was initialised with
   * a public key (`pk_`), `/api/v1/devices` for a secret key (`np_`).
   *
   * Idempotent on `(app_id, token, user_id)`. Returns `created: true`
   * when a new row was inserted, `created: false` when an existing
   * device matched.
   */
  async register(input: RegisterDeviceRequest): Promise<RegisterDeviceResponse> {
    const path = this.http.authScheme === "Public" ? "/api/v1/public/devices" : "/api/v1/devices"
    return await this.http.request<RegisterDeviceResponse>("POST", path, {
      body: toWire(input),
    })
  }

  /**
   * List devices (secret key only). Wraps `GET /api/v1/devices`.
   *
   * Pass `userId` to fetch one end-user's registered devices. The push
   * token is never returned. Returns `{ data, total }`.
   */
  async list(query: ListDevicesQuery = {}): Promise<ListDevicesResponse> {
    const wire: Record<string, string | number | undefined> = {
      user_id: query.userId,
      platform: query.platform,
      status: query.status,
      page: query.page,
      page_size: query.pageSize,
    }

    const raw = await this.http.request<RawListDevicesResponse>("GET", "/api/v1/devices", {
      query: wire,
    })

    return {
      total: raw.total,
      data: raw.data.map((d) => ({
        id: d.id,
        userId: d.user_id,
        platform: d.platform,
        status: d.status,
        tags: d.tags,
        timezone: d.timezone,
        apnsEnvironment: d.apns_environment,
        lastSeenAt: d.last_seen_at,
        insertedAt: d.inserted_at,
      })),
    }
  }

  /**
   * Update a device (e.g. replace its tags). Wraps `PUT /api/v1/devices/:id`.
   *
   * Returns `{ id, tags }`. Throws a `NitropingError` with
   * `code: "not_found"` if the id doesn't belong to your app.
   */
  async update(id: string, input: UpdateDeviceRequest): Promise<UpdateDeviceResponse> {
    const body: Record<string, unknown> = {}
    if (input.tags !== undefined) body["tags"] = input.tags
    return await this.http.request<UpdateDeviceResponse>(
      "PUT",
      `/api/v1/devices/${encodeURIComponent(id)}`,
      { body },
    )
  }

  /**
   * Deactivate a device (soft delete — sets `status = inactive`).
   *
   * Returns `{ id, status: "inactive" }`. Throws a `NitropingError`
   * with `code: "not_found"` if the id doesn't belong to your app.
   */
  async deactivate(id: string): Promise<{ id: string; status: string }> {
    return await this.http.request("DELETE", `/api/v1/devices/${encodeURIComponent(id)}`)
  }

  /**
   * Deactivate a device by its provider token (logout flow — you know the
   * token but not the device id). Wraps `DELETE /api/v1/devices` with a
   * `{ token }` body.
   *
   * Returns `{ id, status: "inactive" }`. Throws a `NitropingError` with
   * `code: "not_found"` when no device with that token belongs to your app.
   */
  async deactivateByToken(token: string): Promise<{ id: string; status: string }> {
    return await this.http.request("DELETE", "/api/v1/devices", { body: { token } })
  }
}

function toWire(input: RegisterDeviceRequest): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    token: input.token,
    platform: input.platform,
  }
  if (input.userId !== undefined) wire["user_id"] = input.userId
  if (input.webPushP256dh !== undefined) wire["web_push_p256dh"] = input.webPushP256dh
  if (input.webPushAuth !== undefined) wire["web_push_auth"] = input.webPushAuth
  if (input.metadata !== undefined) wire["metadata"] = input.metadata
  if (input.tags !== undefined) wire["tags"] = input.tags
  if (input.environment !== undefined) wire["environment"] = input.environment
  if (input.timezone !== undefined) wire["timezone"] = input.timezone
  return wire
}

File: go/devices.go
package nitroping

import (
	"context"
	"errors"
	"net/url"
	"strconv"
)

// DevicesService wraps the /api/v1/devices endpoints. Construct one
// via NewClient.
type DevicesService struct {
	transport *transport
}

// Register registers (or updates) a device with the secret API key.
// Use this for iOS and Android where you control the server. For
// browser flows use the public-key web subscribe API instead (Go SDK
// is server-side only, so there's no equivalent here yet).
//
// The endpoint is idempotent on (app_id, token, user_id) — sending the
// same triple twice returns the existing row with Created=false.
//
//	res, err := client.Devices.Register(ctx, nitroping.DeviceRequest{
//	    Platform: nitroping.PlatformIOS,
//	    Token:    "apns-hex-token",
//	    UserID:   nitroping.String("user-42"),
//	})
func (s *DevicesService) Register(
	ctx context.Context,
	req DeviceRequest,
	opts ...RequestOption,
) (*DeviceResult, error) {
	cfg := applyOptions(opts)
	path := "/api/v1/devices"
	if s.transport.authScheme == "Public" {
		path = "/api/v1/public/devices"
	}
	var out DeviceResult
	if err := s.transport.do(ctx, "POST", path, req, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List lists devices via GET /api/v1/devices (secret key only). The
// optional query filters by user id, platform, status, and paginates.
// The push token is never returned — each row is a DeviceSummary with
// no token field. On success the result is {Data, Total}.
//
//	res, err := client.Devices.List(ctx, nitroping.ListDevicesQuery{
//	    UserID:   "user-42",
//	    Platform: nitroping.PlatformIOS,
//	})
func (s *DevicesService) List(
	ctx context.Context,
	query ListDevicesQuery,
	opts ...RequestOption,
) (*ListDevicesResult, error) {
	cfg := applyOptions(opts)

	q := url.Values{}
	if query.UserID != "" {
		q.Set("user_id", query.UserID)
	}
	if query.Platform != "" {
		q.Set("platform", string(query.Platform))
	}
	if query.Status != "" {
		q.Set("status", query.Status)
	}
	if query.Page != nil {
		q.Set("page", strconv.Itoa(*query.Page))
	}
	if query.PageSize != nil {
		q.Set("page_size", strconv.Itoa(*query.PageSize))
	}

	path := "/api/v1/devices"
	if encoded := q.Encode(); encoded != "" {
		path += "?" + encoded
	}

	var out ListDevicesResult
	if err := s.transport.do(ctx, "GET", path, nil, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update updates a device via PUT /api/v1/devices/:id. Currently the
// only updatable field is the tag set (req.Tags). On success the result
// is {ID, Tags} reflecting the device's tags after the update.
//
// Returns *APIError with Code="not_found" if the id doesn't belong to
// the caller's app.
//
//	res, err := client.Devices.Update(ctx, "dev-1", nitroping.UpdateDeviceRequest{
//	    Tags: []string{"beta", "vip"},
//	})
func (s *DevicesService) Update(
	ctx context.Context,
	id string,
	req UpdateDeviceRequest,
	opts ...RequestOption,
) (*UpdateDeviceResult, error) {
	if id == "" {
		return nil, errors.New("nitroping: device id is required")
	}
	cfg := applyOptions(opts)
	var out UpdateDeviceResult
	path := "/api/v1/devices/" + url.PathEscape(id)
	if err := s.transport.do(ctx, "PUT", path, req, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Deactivate soft-deletes a device (sets status=inactive). Future
// sends skip it. Returns *APIError with Code="not_found" if the id
// doesn't belong to the caller's app.
func (s *DevicesService) Deactivate(
	ctx context.Context,
	id string,
	opts ...RequestOption,
) (*DeactivateResult, error) {
	if id == "" {
		return nil, errors.New("nitroping: device id is required")
	}
	cfg := applyOptions(opts)
	var out DeactivateResult
	path := "/api/v1/devices/" + url.PathEscape(id)
	if err := s.transport.do(ctx, "DELETE", path, nil, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeactivateByToken soft-deletes a device by its provider token rather
// than its id (the logout flow: you hold the push token but not the
// device id). It sends DELETE /api/v1/devices with a {"token": ...}
// body. Returns *APIError with Code="not_found" if no device with that
// token belongs to the caller's app.
//
// This is the sibling of Deactivate, which deletes by id.
func (s *DevicesService) DeactivateByToken(
	ctx context.Context,
	token string,
	opts ...RequestOption,
) (*DeactivateResult, error) {
	if token == "" {
		return nil, errors.New("nitroping: device token is required")
	}
	cfg := applyOptions(opts)
	body := struct {
		Token string `json:"token"`
	}{Token: token}
	var out DeactivateResult
	if err := s.transport.do(ctx, "DELETE", "/api/v1/devices", body, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
